server/service/src/processors/transfer/shipment/update_inbound_shipment.rs

```diff
@@ -0,0 +1,99 @@
+use repository::{
+    InvoiceLineRowRepository, InvoiceRow, InvoiceRowRepository, InvoiceRowStatus, InvoiceRowType,
+    RepositoryError, StorageConnection,
+};
+
+use super::{
+    common::regenerate_inbound_shipment_lines, Operation, ShipmentTransferProcessor,
+    ShipmentTransferProcessorRecord,
+};
+
+const DESCRIPTION: &'static str = "Update inbound shipment from outbound shipment";
+
+pub(crate) struct UpdateInboundShipmentProcessor;
+
+impl ShipmentTransferProcessor for UpdateInboundShipmentProcessor {
+    fn get_description(&self) -> String {
+        DESCRIPTION.to_string()
+    }
+
+    /// Inbound shipment will be updated when all below conditions are met:
+    ///
+    /// 1. Source shipment name_id is for a store that is active on current site (transfer processor driver guarantees this)
+    /// 2. Source shipment is Outbound shipment
+    /// 3. Linked shipment exists (the inbound shipment)
+    /// 4. Linked inbound shipment is Picked (Inbound shipment can only be deleted before it turns to Shipped status)
+    /// 5. Source outbound shipment is Shipped
+    ///
+    /// Only runs once:
+    /// 6. Because linked inbound shipment will be changed to Shipped status and `4.` will never be true again
+    fn try_process_record(
+        &self,
+        connection: &StorageConnection,
+        record_for_processing: &ShipmentTransferProcessorRecord,
+    ) -> Result<Option<String>, RepositoryError> {
+        // Check can execute
+        let (outbound_shipment, linked_shipment) = match &record_for_processing.operation {
+            Operation::Upsert {
+                shipment,
+                linked_shipment,
+            } => (shipment, linked_shipment),
+            _ => return Ok(None),
+        };
+        // 2.
+        if outbound_shipment.invoice_row.r#type != InvoiceRowType::OutboundShipment {
+            return Ok(None);
+        }
+        // 3.
+        let inbound_shipment = match &linked_shipment {
+            Some(linked_shipment) => linked_shipment,
+            None => return Ok(None),
+        };
+        // 4.
+        if inbound_shipment.invoice_row.status != InvoiceRowStatus::Picked {
+            return Ok(None);
+        }
+        // 5.
+        if outbound_shipment.invoice_row.status != InvoiceRowStatus::Shipped {
+            return Ok(None);
+        }
+
+        // Execute
+        let (line_ids_to_delete, new_inbound_lines) = regenerate_inbound_shipment_lines(
```

Comment:
Don't like that the inbound_shipment_lines get a new UUID on every status update.
However I can't think of a nice way to solve this without much more complex code.
As long as nothing links to these uuids I'm sure it's fine.

Reply:
I *think* that this is relying on any new, deleted or updated`outbound shipment lines` to be created/updated before this trigger runs. Is this a safe assumption? What if something triggers this halfway through a sync or something?

Reply:
We've tried pretty hard to guarantee logical data integrity in synchronisation, synced data is integrated inside a transaction (all records are integrated at once). 

And in service layer, there is validation that lines are not editable in shipped status for outbound shipment